src/main/java/com/es/phoneshop/service/impl/CartServiceImpl.java

```diff
@@ -0,0 +1,83 @@
+package com.es.phoneshop.service.impl;
+
+import com.es.phoneshop.exception.OutOfStockException;
+import com.es.phoneshop.model.Cart;
+import com.es.phoneshop.model.CartItem;
+import com.es.phoneshop.model.Product;
+import com.es.phoneshop.service.CartService;
+import com.es.phoneshop.service.ProductService;
+import jakarta.servlet.http.HttpSession;
+
+import java.util.Optional;
+import java.util.UUID;
+
+public class CartServiceImpl implements CartService {
+    private static final String CART_SESSION_ATTRIBUTE = CartServiceImpl.class.getName() + ".cart";
+    private final ProductService productService;
+
+    private CartServiceImpl() {
+        this.productService = new ProductServiceImpl();
+    }
+
+    private static final class SingletonHolder {
+        private static final CartServiceImpl INSTANCE = new CartServiceImpl();
+    }
+
+    public static CartServiceImpl getInstance() {
+        return SingletonHolder.INSTANCE;
+    }
+
+    @Override
+    public Cart get(HttpSession session) {
+        synchronized (session.getId()) {
+            Cart cart = (Cart) session.getAttribute(CART_SESSION_ATTRIBUTE);
+            if (cart == null) {
+                cart = new Cart();
+                session.setAttribute(CART_SESSION_ATTRIBUTE, cart);
+            }
+
+            return cart;
+        }
+    }
+
+    @Override
+    public void add(UUID productId, int quantity, HttpSession session) throws OutOfStockException {
+        Product product = productService.getProduct(productId);
+        Cart cart = this.get(session);
+
+        synchronized (session.getId()) {
```

Comment:
please read this article
https://www.baeldung.com/java-synchronization-bad-practices